OWML.ModHelper.Input/ModInputHandler.cs

```diff
@@ -0,0 +1,338 @@
+        {
+            return (!(IsPressedExact(combination) || IsPressedSingle(combination)))
+                && (combination.PressDuration < TapDuration) 
+                && combination.IsFirst(keep);
+        }
+
+        public bool WasNewlyReleased(IModInputCombination combination, bool keep = false)
+        {
+            return (!(IsPressedExact(combination) || IsPressedSingle(combination)))
+                && combination.IsFirst(keep);
+        }
+
+        private KeyCode StringToKeyCode(string key)
+        {
+            key = key.Trim();
+            if (key.Contains(XboxPrefix))
+            {
+                var xboxKey = key.Substring(XboxPrefix.Length);
+                var xboxCode = (XboxButton)Enum.Parse(typeof(XboxButton), xboxKey, true);
+                if (Enum.IsDefined(typeof(XboxButton), xboxCode))
+                {
+                    return InputTranslator.GetKeyCode(xboxCode, false);
+                }
+                return KeyCode.None;
+            }
+            var changedKey = key;
+            if (key == "control" || key == "ctrl")
+            {
+                return KeyCode.LeftControl;
+            }
+            if (key == "shift")
+            {
+                return KeyCode.LeftShift;
+            }
+            if (key == "alt")
+            {
+                return KeyCode.LeftAlt;
+            }
+            var code = (KeyCode)Enum.Parse(typeof(KeyCode), changedKey, true);
+            if (Enum.IsDefined(typeof(KeyCode), code))
+            {
+                return code;
+            }
+            return KeyCode.None;
+        }
+
+        private long ParseCombination(string combo, bool forRemoval = false)
+        {
+            var thisCombination = new int[MaxComboLength];
+            var i = 0;
+            foreach (var key in combo.Trim().ToLower().Split('+'))
+            {
+                if (i >= MaxComboLength)
+                {
+                    return (int)RegistrationCode.CombinationTooLong;
+                }
+                KeyCode code = StringToKeyCode(key);
+                if (code == KeyCode.None)
+                {
+                    return (int)RegistrationCode.InvalidCombination;
+                }
+                thisCombination[i] = (int)code;
+                i++;
+            }
+            Array.Sort(thisCombination);
+            long hash = 0;
+            for (i = 0; i < MaxComboLength; i++)
+            {
+                hash = hash * MaxUsefulKey + thisCombination[i];
+            }
+            if (hash < MaxUsefulKey && _gameBindingCounter[hash] > 0)
+            {
+                return (int)RegistrationCode.CombinationTaken;
+            }
+            return (_comboRegistry.ContainsKey(hash) && !forRemoval ? (int)RegistrationCode.CombinationTaken : hash);
+        }
+
+        private RegistrationCode SwapCombination(ModInputCombination combination, bool toUnregister)
```

Comment:
This method is a bit complicated to read. I think it would be nice if you extract the contents into smaller methods with intuitive names.

Reply:
separated part that generated list of hashes, not sure if that makes sense to separate farther